test/gracefulstop_test.go

```diff
@@ -216,3 +217,90 @@ func (s) TestGracefulStopClosesConnAfterLastStream(t *testing.T) {
 		<-gracefulStopDone // Wait for GracefulStop to return.
 	})
 }
+
+func (s) TestGracefulStopBlocksUntilGRPCConnectionsTerminate(t *testing.T) {
+	// This tests ensures that GracefulStop() blocks until all ongoing
+	// client grpc-calls finished.
+	unblockGRPCCall := make(chan struct{})
+	grpcCallExecuting := make(chan struct{})
+	ss := &stubserver.StubServer{
+		UnaryCallF: func(ctx context.Context, in *testpb.SimpleRequest) (*testpb.SimpleResponse, error) {
+			close(grpcCallExecuting)
+			<-unblockGRPCCall
+			return &testpb.SimpleResponse{}, nil
+		},
+	}
+
+	err := ss.Start(nil)
+	if err != nil {
+		t.Fatalf("StubServer.start failed: %s", err)
+	}
+	t.Cleanup(ss.Stop)
+
+	grpcClientCallReturned := make(chan struct{})
+	go func() {
+		clt := ss.Client
+		_, err := clt.UnaryCall(context.Background(), &testpb.SimpleRequest{})
+		if err != nil {
+			t.Errorf("grpc call failed with error: %s", err)
+		}
+		close(grpcClientCallReturned)
+	}()
+
+	gracefulStopReturned := make(chan struct{})
+	<-grpcCallExecuting
+	go func() {
+		ss.S.GracefulStop()
+		close(gracefulStopReturned)
+	}()
+
+	time.Sleep(time.Second)
+	select {
+	case <-gracefulStopReturned:
+		t.Error("GracefulStop returned before GRPC method call ended")
+	default:
+	}
+
+	unblockGRPCCall <- struct{}{}
+	<-grpcClientCallReturned
+	<-gracefulStopReturned
+}
+
+func (s) TestStopAbortsBlockingGAbortsBlockingGRPCCall(t *testing.T) {
```

Comment:
The test for `Stop()` is in a file called `GracefulStop()`.
Creating an extra file `stop_test.go` file for a single tests seems over the top me.

We could:
- keep  the Stop() testcase in the gracefulstop_test.go file for now
- rename the `gfracefulstop_test.go` file to `server_stop_test.go`
- move all tests from gracefulstop_test.go to server_test.go

Reply:
My preference would be to move these tests to the root directory and put them in the `grpc_test` package.  Directly in `server_ext_test.go` seems perfect.

If for some reason that doesn't work (like you need functionality that's in the `test/` package), then that's fine, you can leave them there.

Also, no need to move the existing tests (unless you want to and it turns out to be easy, in which case SGTM).

Reply:
Ok, then I'll keep the tests where they are.
They use `isConnClosedErr()` and now also `defaultTestShortTimeout` from the test package.